Fix RESOURCE_EXHAUSTED error when dealing with large datasets in Flax example scripts

duongna21 opened the pull request:
# What does this PR do?

Fixes #15411. All Flax example scripts using `generate_batch_splits` raise RESOURCE_EXHAUSTED error when training with large datasets on TPU.

## Fix

Simply transfer the jnp array `samples_idx` back to the host before putting it in `np.split`.

## Who can review?
cc potential reviewers: @sgugger, @patil-suraj, @patrickvonplaten

duongna21 committed:
Fix RESOURCE_EXHAUSTED error for large datasets on Flax example scripts

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18069). All of your documentation changes will be reflected on that endpoint.

duongna21 commented:
@ydshieh I dug a little bit into the memory profiling when running the original code:

```
>>> import jax.numpy as jnp
>>> import numpy as np
>>> import time
>>> def generate_batch_splits(samples_idx: jnp.ndarray, batch_size: int) -> jnp.ndarray:
...     num_samples = len(samples_idx)
...     samples_to_remove = num_samples % batch_size
...     if samples_to_remove != 0:
...         samples_idx = samples_idx[:-samples_to_remove]
...     sections_split = num_samples // batch_size
...     batch_idx = np.split(samples_idx, sections_split)
...     jax.profiler.save_device_memory_profile("generate_batch_splits_tpu_bs64.prof")
...     return batch_idx
...  
>>> input_rng = jax.random.PRNGKey(0)
>>> samples_idx = jax.random.permutation(input_rng, jnp.arange(50758958)) # 50758958 is num of samples in my case
>>> tic = time.time()
>>> train_batch_idx = generate_batch_splits(samples_idx, batch_size=64) # per device batch size = 8
>>> toc = time.time()
>>> print('elapsed time: ', toc-tic)
elapsed time:  578.8218853473663
```

<img src="https://i.imgur.com/8tns1X2.png" width="400" height="300" />


=> **968.15MB** on TPU memory wasted while generating batch splits. When I transfered the jnp array `samples_idx` back to the CPU before putting it in `np.split`:
```
...
def generate_batch_splits(samples_idx: jnp.ndarray, batch_size: int) -> jnp.ndarray:
    samples_idx = jax.device_get(samples_idx)
    num_samples = len(samples_idx)
    ...
>>> print('elapsed time: ', toc-tic)
elapsed time:  3.2475574016571045
```
<img src="https://i.imgur.com/MCDDYHv.png" width="220" height="300" />

=> We just saved **968.15MB** (and 575 seconds). But we have an even better solution to save another **193.63MB** (and 2 seconds) by generating `samples_idx` using CPU in the first place:

```
...
samples_idx = np.random.permutation(np.arange(50758958))  # 50758958 is num of samples in my case
...
>>> print('elapsed time: ', toc-tic)
elapsed time:  1.0959625244140625
```

Anyway, 1GB of wasted TPU memory is not the decisive factor in causing RESOURCE_EXHAUTED, but this fix saves us from the weird RESOURCE_EXHAUTED error when using the same batch size and only increasing the dataset size.

I pushed another commit with the comment.

duongna21 committed:
using np.permutation for creating batch_idx

duongna21 committed:
train_samples_idx -> training_samples_idx

duongna21 committed:
fix type hints

patrickvonplaten commented:
Thanks a lot for the PR @duongna21 !
You're right we should better use numpy there - the PR looks good to me :-)

patrickvonplaten reviewed (approved):
It'll break reproducibility to some extent, but this should be fine - IMO it's more important that it doesn't fail on TPU
